### PR TITLE
WhatWeb agent

### PR DESCRIPTION
**Pull request by RabsonJ**
Alpha implementation of whatweb agent

### CODE REVIEW

File: agent/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+"""Init file for the agent."""
```
Comment (deadly-panda) [inline agent/__init__.py:1]: I think u can remove this

File: agent/whatweb.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
+"""WhatWeb Agent : Agent responsible for finger-printing a website."""
+import logging
+import json
+import subprocess
+import tempfile
+import os
+from typing import Union
+from rich import logging as rich_logging
```
Comment (deadly-panda) [inline agent/whatweb.py:8]: this line should move to the second group of imports

File: agent/whatweb.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+from ostorlab.agent import agent
+from ostorlab.agent import message as msg
+
+from agent import definitions as whatweb_definitions
+
+logging.basicConfig(
+    format='%(message)s',
+    datefmt='[%X]',
+    handlers=[rich_logging.RichHandler(rich_tracebacks=True)]
+)
```
Comment (deadly-panda) [inline agent/whatweb.py:19]: I think you should add : level='INFO', force=True
But please confirm with Alae or Amine.

File: Dockerfile [hunk 4]
```diff
@@ -20,4 +20,4 @@ ENV PYTHONPATH=/app
 COPY agent /app/agent
 COPY ostorlab.yaml /app/agent/ostorlab.yaml
 WORKDIR /app
-CMD ["python3", "/app/agent/whatweb_agent.py"]
+CMD ["python3", "/app/agent/whatweb.py"]
```
Comment (3asm) [inline Dockerfile:23]: Why was it renamed to whatweb only, all the other agents, the main agent class has the name agent in it.

File: agent/__init__.py (hunk 1 above)
Comment (3asm) [inline agent/__init__.py:1]: Remove comment.

File: agent/definitions.py [hunk 5]
```diff
@@ -0,0 +1,9 @@
+"""Data types used by the whatweb agent."""
+
+from dataclasses import dataclass
+
+@dataclass
+class Target:
```
Comment (3asm) [inline agent/definitions.py]: This dataclass sounds like an overkill.

File: agent/whatweb.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
+"""WhatWeb Agent : Agent responsible for finger-printing a website."""
```
Comment (3asm) [inline agent/whatweb.py:1]: I think this should be renamed to agent_whatweb?

File: tests/whatweb_test.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+"""Unittests for whatweb agent."""
+
+import tempfile
+
+from ostorlab.agent import message as msg
+
+
+def testWhatWebAgent_allChecks_emitsFingerprints(whatweb_test_agent, mocker):
```
Comment (3asm) [inline tests/whatweb_test.py:8]: The 
```suggestion
def testWhatWebAgent_withAllChecksEnabled_emitsFingerprints(whatweb_test_agent, mocker):
```

File: tests/whatweb_test.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    """Test the whatweb agent with a given target address"""
```
Comment (3asm) [inline tests/whatweb_test.py:9]: Please more details for the unit test.

File: agent/whatweb.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+    def process(self, message: msg.Message) -> None:
+        """Starts a whatweb scan, wait for the scan to finish,
+        and emit the results.
+
+        Args:
+            message:  The message to process from ostorlab runtime.
+        """
+        logger.info('processing message of selector : %s', message.selector)
+        self._scan(domain_name=message.data['name'])
```
Comment (3asm) [inline agent/whatweb.py:130]: Please inline the _scan method, adding an extra method here is not useful.

File: agent/whatweb.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    def _scan(self, domain_name: str):
+        """Start a scan, wait for the scan results and clean the scan output.
+
+           returns:
+            - Scan results from whatweb.
+        """
+        with tempfile.TemporaryFile() as fp:
+            self._start_scan(domain_name, fp.name)
+            fp.seek(0)
+            self._parse_result(domain_name, fp)
```
Comment (3asm) [inline agent/whatweb.py:120]: This seems to parse and emit, please rename method or separate the method.

File: agent/whatweb.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+WHATWEB_PATH = './whatweb'
+WHATWEB_DIRECTORY = '/WhatWeb'
+SELECTOR = 'v3.fingerprint.domain_name.library'
+
+
+class WhatWebAgent(agent.Agent):
+    """Agent responsible for finger-printing a website."""
+
+    def _start_scan(self, domain_name: str, output_file: Union[str, bytes, os.PathLike]):
```
Comment (3asm) [inline agent/whatweb.py:43]: Please reorder the method in the order they are called for easier readability.

File: agent/whatweb.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+                    for result in results:
+                        if isinstance(result, list):
+                            for list_plugin in result:
+                                if len(list_plugin) > 0:
+                                    plugin = list_plugin[0]
+                                else:
+                                    plugin = list_plugin
+                                if plugin not in BLACKLISTED_PLUGINS:
+                                    values = list_plugin[1]
+                                    versions = ''
```
Comment (3asm) [inline agent/whatweb.py:71]: Why is versions defaulting to an empty string?

File: agent/whatweb.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+                                        if 'version' in value:
+                                            versions = value['version']
+                                        if 'string' in value:
+                                            name = str(value['string'])
+                                    self._send_detected_fingerprints(domain_name, name, versions)
+                logger.info('Scan is done Parsing the results from %s.', output_file.name)
+        except OSError as e:
+            logger.error('Exception while processing %s with message %s', output_file, e)
+
+    def _send_detected_fingerprints(self, domain_name: str, name: str, versions: Union[list, str]):
```
Comment (3asm) [inline agent/whatweb.py:83]: This method should always get passed versins as a list. If we have only, put it in a list.

File: agent/whatweb.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+        """Emits the identified fingerprints.
+
+        Args:
+            domain_name: The domain name.
+            name: The name of the website.
+            versions: The versions identified by WhatWeb Agent
+        """
+
+        fingerprint_type = FINGERPRINT_TYPE[name.lower()] if name.lower() in FINGERPRINT_TYPE else DEFAULT_FINGERPRINT
+        if isinstance(versions, list):
```
Comment (3asm) [inline agent/whatweb.py:93]: This code should only assume the list case.

**Review by 3asm — CHANGES_REQUESTED** [review]

**Review by deadly-panda — CHANGES_REQUESTED** [review]

**Review by 3asm — APPROVED** [review]

**Pull request merged**